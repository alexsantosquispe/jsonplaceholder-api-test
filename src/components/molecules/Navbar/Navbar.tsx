import cn from 'clsx';
import { useState } from 'react';
import { twMerge } from 'tailwind-merge';

import { MenuButton } from './components/MenuButton';
import { NavbarLogo } from './components/NavbarLogo';
import { NavbarMenu } from './components/NavbarMenu';
import { SwitchThemeButton } from '../../atoms/SwitchThemeButton/SwitchThemeButton';

export const Navbar = () => {
  const [isMenuOpen, setIsMenuOpen] = useState(false);

  const toggleMenu = () => {
    setIsMenuOpen((prev) => !prev);
  };

  const closeMenu = () => {
    setIsMenuOpen(false);
  };

  return (
    <header className="fixed z-40 flex w-full justify-center border-b border-b-gray-100 bg-white/60 backdrop-blur-sm md:h-[4rem] dark:border-b-white/10 dark:bg-black/50">
      <nav className="flex w-full max-w-[80rem] flex-col items-center justify-between md:flex-row md:px-6">
        <div
          className={twMerge(
            'flex w-full items-center justify-between py-1 md:border-none',
            cn({
              'border-b border-b-gray-100 dark:border-b-white/10': isMenuOpen
            })
          )}
        >
          <MenuButton isOpen={isMenuOpen} onClick={toggleMenu} />
          <NavbarLogo />
          <SwitchThemeButton classNameContainer="md:hidden" />
        </div>

        <NavbarMenu isMenuOpen={isMenuOpen} onSelectOption={closeMenu} />
      </nav>
    </header>
  );
};
