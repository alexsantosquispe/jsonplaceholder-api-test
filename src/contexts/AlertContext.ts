import { createContext } from 'react';

import { ALERT_TYPES } from '../constants';

export type AlertType = {
  id: string;
  title: string;
  message: string;
  type?: (typeof ALERT_TYPES)[keyof typeof ALERT_TYPES];
  timeout?: number;
};

interface AlertContextType {
  alerts: AlertType[];
  addAlert: (alert: AlertType) => void;
  removeAlert: (alertId: string) => void;
}

const AlertContext = createContext<AlertContextType | null>(null);

export default AlertContext;
