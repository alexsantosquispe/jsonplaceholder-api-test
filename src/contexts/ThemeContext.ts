import { createContext } from 'react';

export const Theme = {
  light: 'light',
  dark: 'dark'
} as const;

export type ThemeType = keyof typeof Theme;

interface ThemeContextType {
  theme: ThemeType;
  toggleTheme: () => void;
}

const ThemeContext = createContext<ThemeContextType | null>(null);

export default ThemeContext;
