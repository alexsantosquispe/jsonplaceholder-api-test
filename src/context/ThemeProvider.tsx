import { ReactNode, useState } from 'react';

import ThemeContext, { ThemeType } from './ThemeContext';

interface ThemeProviderProps {
  children: ReactNode;
}

export const ThemeProvider = ({ children }: ThemeProviderProps) => {
  const [theme, setTheme] = useState<ThemeType>('light');
  const toggleTheme = () => {
    setTheme((prev) => (prev === 'light' ? 'dark' : 'light'));
  };

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};
