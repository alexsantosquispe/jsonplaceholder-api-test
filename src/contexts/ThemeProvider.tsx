import { ReactNode, useEffect, useState } from 'react';

import ThemeContext, { Theme, ThemeType } from './ThemeContext';
import { THEME_KEY } from '../constants';
import {
  applyThemeToDocument,
  getStoredTheme,
  getSystemTheme,
  setLSValue
} from '../utils';

interface Props {
  children: ReactNode;
}

export const ThemeProvider = ({ children }: Props) => {
  const initialTheme = getStoredTheme() ?? getSystemTheme();
  const [currentTheme, setCurrentTheme] = useState<ThemeType>(initialTheme);

  useEffect(() => {
    applyThemeToDocument(currentTheme);
    setLSValue(THEME_KEY, currentTheme);
  }, [currentTheme]);

  useEffect(() => {
    const storedTheme = getStoredTheme();
    if (storedTheme) return;

    const mediaQuery = window.matchMedia('(prefers-color-scheme: dark)');

    const handleChange = (e: MediaQueryListEvent) => {
      setCurrentTheme(e.matches ? Theme.dark : Theme.light);
    };

    mediaQuery.addEventListener('change', handleChange);
    return () => mediaQuery.removeEventListener('change', handleChange);
  }, []);

  const toggleTheme = () => {
    setCurrentTheme((prev) => (prev === Theme.dark ? Theme.light : Theme.dark));
  };

  return (
    <ThemeContext.Provider value={{ theme: currentTheme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};
