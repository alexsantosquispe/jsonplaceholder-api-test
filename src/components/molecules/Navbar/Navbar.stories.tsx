import { Meta, StoryObj } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

import { Navbar } from './Navbar';
import { ThemeProvider } from '../../../contexts/ThemeProvider';

const meta = {
  title: 'Molecules/Navbar',
  component: Navbar,
  parameters: {
    layout: 'fullscreen'
  }
} satisfies Meta<typeof Navbar>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = () => {
  return (
    <ThemeProvider>
      <BrowserRouter
        future={{
          v7_startTransition: true,
          v7_relativeSplatPath: true
        }}
      >
        <div className="dark:bg-primary/90 flex w-full flex-col bg-white">
          <Navbar />
          <main className="dark:bg-primary/90 mt-14 flex w-full grow justify-center border-b border-gray-200 bg-gray-50 md:p-0">
            <div className="text-primary flex h-[30rem] items-center justify-center dark:text-white">
              Main content
            </div>
          </main>
        </div>
      </BrowserRouter>
    </ThemeProvider>
  );
};

Default.args = {};
