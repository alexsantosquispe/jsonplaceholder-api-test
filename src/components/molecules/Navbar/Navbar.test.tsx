import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import { Navbar } from './Navbar';
import { ThemeProvider } from '../../../contexts/ThemeProvider';

Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn()
  }))
});

describe('Navbar', () => {
  it('should render the component correctly', () => {
    const component = render(
      <ThemeProvider>
        <BrowserRouter
          future={{
            v7_startTransition: true,
            v7_relativeSplatPath: true
          }}
        >
          <Navbar />
        </BrowserRouter>
      </ThemeProvider>
    );

    expect(component).toMatchSnapshot();
  });

  it('should display the nav items', () => {
    render(
      <ThemeProvider>
        <BrowserRouter
          future={{
            v7_startTransition: true,
            v7_relativeSplatPath: true
          }}
        >
          <Navbar />
        </BrowserRouter>
      </ThemeProvider>
    );

    expect(screen.getByText('JSON')).toBeInTheDocument();
    expect(screen.getByText('Placeholder')).toBeInTheDocument();
    expect(screen.getByText('Posts')).toBeInTheDocument();
    expect(screen.getByText('Todos')).toBeInTheDocument();
    expect(screen.getByTestId('moon-icon')).toBeInTheDocument();
  });
});
