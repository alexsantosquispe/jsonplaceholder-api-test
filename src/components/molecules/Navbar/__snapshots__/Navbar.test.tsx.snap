// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar should render the component correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="fixed z-40 flex w-full justify-center border-b border-b-gray-100 bg-white/70 backdrop-blur-sm md:h-14 dark:border-b-white/10 dark:bg-black/50"
      >
        <nav
          class="flex w-full max-w-[80rem] flex-col items-center justify-between md:flex-row md:px-6"
        >
          <div
            class="flex w-full items-center justify-between py-1 md:border-none"
          >
            <a
              href="/"
            >
              <div
                class="text-primary flex align-middle text-xl leading-4 md:p-0 md:text-2xl md:leading-9 dark:text-white/90"
              >
                <p
                  class="font-light"
                >
                  <span
                    class="font-bold"
                  >
                    { JSON }
                  </span>
                   Placeholder
                  <span
                    class="text-accent text-5xl leading-1 font-bold dark:text-orange-400"
                  >
                    .
                  </span>
                </p>
              </div>
            </a>
          </div>
          <div
            class="flex w-full flex-col items-start overflow-hidden text-[0.9375rem] md:flex-row md:items-center md:justify-end md:gap-x-2 md:text-black/50 dark:text-white/60 h-0 md:h-auto"
          >
            <a
              class="w-full py-3 px-4 md:w-fit hover:text-primary dark:hover:text-white"
              href="/posts"
            >
              Posts
            </a>
            <a
              class="w-full py-3 px-4 md:w-fit hover:text-primary dark:hover:text-white"
              href="/todos"
            >
              Todos
            </a>
            <button
              aria-label="dark mode icon"
              class="transition-color cursor-pointer rounded-md p-4 text-black duration-150 ease-in-out hover:bg-gray-100 md:p-2 dark:text-white dark:hover:bg-white/10 hidden md:block"
            >
              <svg
                class="size-4"
                data-testid="moon-icon"
                fill="none"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M21.752 15.002A9.72 9.72 0 0 1 18 15.75c-5.385 0-9.75-4.365-9.75-9.75 0-1.33.266-2.597.748-3.752A9.753 9.753 0 0 0 3 11.25C3 16.635 7.365 21 12.75 21a9.753 9.753 0 0 0 9.002-5.998Z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </button>
          </div>
        </nav>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="fixed z-40 flex w-full justify-center border-b border-b-gray-100 bg-white/70 backdrop-blur-sm md:h-14 dark:border-b-white/10 dark:bg-black/50"
    >
      <nav
        class="flex w-full max-w-[80rem] flex-col items-center justify-between md:flex-row md:px-6"
      >
        <div
          class="flex w-full items-center justify-between py-1 md:border-none"
        >
          <a
            href="/"
          >
            <div
              class="text-primary flex align-middle text-xl leading-4 md:p-0 md:text-2xl md:leading-9 dark:text-white/90"
            >
              <p
                class="font-light"
              >
                <span
                  class="font-bold"
                >
                  { JSON }
                </span>
                 Placeholder
                <span
                  class="text-accent text-5xl leading-1 font-bold dark:text-orange-400"
                >
                  .
                </span>
              </p>
            </div>
          </a>
        </div>
        <div
          class="flex w-full flex-col items-start overflow-hidden text-[0.9375rem] md:flex-row md:items-center md:justify-end md:gap-x-2 md:text-black/50 dark:text-white/60 h-0 md:h-auto"
        >
          <a
            class="w-full py-3 px-4 md:w-fit hover:text-primary dark:hover:text-white"
            href="/posts"
          >
            Posts
          </a>
          <a
            class="w-full py-3 px-4 md:w-fit hover:text-primary dark:hover:text-white"
            href="/todos"
          >
            Todos
          </a>
          <button
            aria-label="dark mode icon"
            class="transition-color cursor-pointer rounded-md p-4 text-black duration-150 ease-in-out hover:bg-gray-100 md:p-2 dark:text-white dark:hover:bg-white/10 hidden md:block"
          >
            <svg
              class="size-4"
              data-testid="moon-icon"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21.752 15.002A9.72 9.72 0 0 1 18 15.75c-5.385 0-9.75-4.365-9.75-9.75 0-1.33.266-2.597.748-3.752A9.753 9.753 0 0 0 3 11.25C3 16.635 7.365 21 12.75 21a9.753 9.753 0 0 0 9.002-5.998Z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
        </div>
      </nav>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
